### PR TITLE
Add OpenCover parser to xml_handler

### PR DESCRIPTION
**Pull request by josephsnyder**
Add a parser which is able to take in a directory of C# code with
an XML result file from OpenCover and parse the coverage correctly.

The handler will also need to read in the source files and set the
initial lines of executable code.

### CODE REVIEW

File: tests/test_opencovercoverage.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+//
+// After including cdash_test_case.php, subsequent require_once calls are
+// relative to the top of the CDash source tree
+//
+require_once(dirname(__FILE__).'/cdash_test_case.php');
+
+class JSCoverCoverageTestCase extends KWWebTestCase
+{
+  function __construct()
+    {
```
Comment (danlamanna) [inline tests/test_opencovercoverage.php]: The code style doesn't conform to the rest of CDash, not sure why the style checker isn't catching it.

File: tests/test_opencovercoverage.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    if(strpos($content, '47.37') === false)
+      {
+      $this->fail('\"47.37\" not found when expected');
+      return 1;
+      }
+
+    return 0;
+    }
+}
+?>
```
Comment (danlamanna) [inline tests/test_opencovercoverage.php]: No need for the ending tag

Comment (danlamanna) [inline tests/test_opencovercoverage.php]: The test should clean up anything it creates in some form of `teardown` method.

File: xml_handlers/OpenCoverTar_handler.php [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        if (!array_key_exists($path, $this->CoverageFileLogs)) {
+            $coverageFileLog = new CoverageFileLog();
+            $coverageFileLog->BuildId =  $this->Build->Id;
+            $coverageFile = new CoverageFile();
+            $coverageFile->FullPath = trim($path);
+            // Get the ID for this coverage file, or create a new empty one
+            //if it doesn't already exist.
+            $sql = pdo_query(
+            "SELECT id FROM coveragefile
+            WHERE fullpath='$path' AND file IS NULL");
```
Comment (danlamanna) [inline xml_handlers/OpenCoverTar_handler.php]: I lose track of how our pdo implementation works, but is this injecting a filename directly into SQL?

File: xml_handlers/OpenCoverTar_handler.php [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+       $inlongComment= false;
+       if ($this->coverageFile) {
+           foreach($fileContents as $key=>$line) {
+              // Matches the beginning of a comment block
+              if(preg_match("/\/[*]+/", trim($line))) {
+                  $inlongComment=true;
+              }
+
+              $this->coverageFile->File .= rtrim($line).'<br>';
+              if((preg_match("/^\/\//", trim($line))) or
```
Comment (danlamanna) [inline xml_handlers/OpenCoverTar_handler.php]: As a matter of preference this logic should probably be flipped so we don't have empty code blocks. 

Also is regex parsing how this is actually done, can this be generated in a more easily parse-able format? If we're trying to glean information about PHP code, is this something we could use an AST parser, or even PHPs builtin reflection API for?

Nitpick: you should just `$line = trim($line)` at the top of all of these.

**Comment by josephsnyder** [comment]
Thanks for the comments, Dan.  I'll get started on the fixes.

File: xml_handlers/OpenCoverTar_handler.php (hunk 4 above)
Comment (josephsnyder) [inline xml_handlers/OpenCoverTar_handler.php]: The regex parsing here is to capture whether or not the **C#** lines that we're reading in the source files are executable or not.  I'll reconfigure it to get rid of the empty block and get rid of the extra "trim" calls.

**Comment by josephsnyder** [comment]
@danlamanna, I think that I've fixed up everything that you mentioned.  PTAL again when you can.

File: xml_handlers/OpenCoverTar_handler.php [hunk 5]
```diff
@@ -231,29 +231,28 @@ public function readSourceFile($buildid, $fileinfo) {
...
-                (preg_match("/^private/", trim($line))) or
-                (preg_match("/^\[/", trim($line))) or
-                (preg_match("/[{}]/", trim($line))) or
-                ("" == trim($line)) or
-                ($inlongComment)) {
-                // ignore these lines
-              } else {
-                  // Assumed to be executable code
+              $this->coverageFile->File .= $displayLine.'<br>';
+              if(!((preg_match("/^\/\//", $trimmedLine)) or
```
Comment (danlamanna) [inline xml_handlers/OpenCoverTar_handler.php]: Shouldn't this negative the entire condition, not just the first?

Comment (josephsnyder) [inline xml_handlers/OpenCoverTar_handler.php]: There is a new set of parentheses there which make it negate the whole conditional not just the first one entry.  It just doesn't quite show in the diff.  I can redo the block to make that a little clearer if you want

Comment (danlamanna) [inline xml_handlers/OpenCoverTar_handler.php]: Ah I see now, my mistake.

**Comment by danlamanna** [comment]
Oh, this isn't being tested against Circle or stylecheck since it's based on a fork, any ideas on how to go about fixing that @zackgalbreath?

**Comment by zackgalbreath** [comment]
> Oh, this isn't being tested against Circle or stylecheck since it's based on a fork, any ideas on how to go about fixing that @zackgalbreath?

Good catch.  I just tweaked Circle so that it will tests PRs from a fork.  It also looks like StyleCI was just plain broken?  Should be fixed now.  I'll make sure on another open PR.

File: xml_handlers/OpenCoverTar_handler.php [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+        $this->Coverages = array();
+        $this->CoverageFiles = array();
+        $this->CoverageFileLogs = array();
+    }
+
+    public function startElement($parser, $name, $attributes) {
+        parent::startElement($parser, $name, $attributes);
+        /**
+         *  SEQUENCEPOINT denotes a line in the source file that is exectuable and
```
Comment (zackgalbreath) [inline xml_handlers/OpenCoverTar_handler.php]: typo in comment (exectuable)

**Pull request merged**